DRILL-4514 : Add describe schema <schema_name> command

arina-ielchiieva opened the pull request:
Syntax:
describe database <db_name>
describe schema <schema_name>

Output:
Command: describe schema dfs;
+--------+------------+
| schema | properties |
+--------+------------+
| dfs | {
  "type" : "file",
  "enabled" : true,
  "connection" : "file:///",
  "config" : null,
  "formats" : {
    "psv" : {
      "type" : "text",
      "extensions" : [ "tbl" ],
      "delimiter" : "|"
    },
    "csv" : {
      "type" : "text",
      "extensions" : [ "csv" ],
      "delimiter" : ","
    },
    "tsv" : {
      "type" : "text",
      "extensions" : [ "tsv" ],
      "delimiter" : "\t"
    },
    "parquet" : {
      "type" : "parquet"
    },
    "json" : {
      "type" : "json",
      "extensions" : [ "json" ]
    },
    "avro" : {
      "type" : "avro"
    },
    "sequencefile" : {
      "type" : "sequencefile",
      "extensions" : [ "seq" ]
    },
    "csvh" : {
      "type" : "text",
      "extensions" : [ "csvh" ],
      "extractHeader" : true,
      "delimiter" : ","
    }
  },
  "location" : "/",
  "writable" : false,
  "defaultInputFormat" : null
} |
+--------+------------+

hnfgns commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DescribeSchemaCommandResult.java:
```diff
@@ -0,0 +1,30 @@
+
+public class DescribeSchemaCommandResult {
+
+  public String name;
```
You may want to make these final.

arina-ielchiieva replied:
Done.

hnfgns commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/parser/SqlDescribeSchema.java:
```diff
@@ -0,0 +1,82 @@
+  public SqlDescribeSchema(SqlParserPos pos, SqlIdentifier schema) {
+    super(pos);
+    this.schema = schema;
+    assert schema != null;
```
What about a precondition here?

arina-ielchiieva replied:
Not sure what you mean. Could you please expand on your comment?

hnfgns replied:
Sure. I meant we should likely use an explicit null-checking pattern like `this.schema = Preconditions.checkNotNull(schema, "msg")` instead of an assertion here. 

ps: Guava's Preconditions.

arina-ielchiieva replied:
Agree, preconditions is better. But I have checked that schema can't come as null, so I have removed check for not null at all.

jacques-n commented on exec/java-exec/src/main/codegen/includes/parserImpls.ftl:
```diff
@@ -278,3 +278,19 @@ SqlNode SqlRefreshMetadata() :
     }
 }
 
+/**
```
What do you think about including this syntax directly in Calcite? @julianhyde, do you think this is generic enough to be shared or should we keep only in Drill code?

arina-ielchiieva replied:
I would say, it's rather specific syntax.

julianhyde commented:
Yes, I think this would be good in Calcite. Thanks for offering.

DESCRIBE is in the SQL standard (the latest draft, anyway; I didn't check any others) with a slightly different purpose -- to describe the columns of a query. (It reminds me a bit of MySQL, where DESCRIBE and EXPLAIN are synonyms.) That's good news, because it means we're not adding another reserved word, but we also need to make sure that DESCRIBE TABLE, DESCRIBE SCHEMA, DESCRIBE DATABASE are compatible with the standard syntax.

```
20.9 <describe statement>

This Subclause is modified by Subclause 17.4, “<describe statement>”, in ISO/IEC 9075-9.

Function

Obtain information about the <select list> columns or
<dynamic parameter specification>s contained in a prepared statement
or about the columns of the result set associated with a cursor.

Format

<describe statement> ::=
    <describe input statement>
  | <describe output statement>
<describe input statement> ::=
    DESCRIBE INPUT <SQL statement name> <using descriptor> [ <nesting option> ]
<describe output statement> ::=
    DESCRIBE [ OUTPUT ] <described object> <using descriptor> [ <nesting option> ]
<nesting option> ::=
    WITH NESTING
  | WITHOUT NESTING
<using descriptor> ::=
    USING [ SQL ] DESCRIPTOR <descriptor name>
<described object> ::=
    <SQL statement name>
    | CURSOR <cursor name> STRUCTURE
```

The output for DESCRIBE TABLE should be the same as Jdbc getTables. Do you agree?

arina-ielchiieva commented:
Created Calcite Jira - https://issues.apache.org/jira/browse/CALCITE-1168

arina-ielchiieva commented:
Comment from Jacques - Let's open a follow-up bug to move this to Calcite and get in Drill for now (was accidentally added to https://github.com/apache/drill/pull/448)

arina-ielchiieva commented:
@hnfgns, @jacques-n 
Could you please review this pull request?

jinfengni commented:
The patch will show "schema_name, location" for "describe schema xxx" when it applied to file system schema. I'm not sure why you only want to make it work for file system schema, and return "information not available" for other schema. Also, what's the reason to only show "location"? What if we add support for other schemas, and will "location" still make sense for non-filesystem schema?

To me, it would make more sense to return all the properties of schema for "describe schema/database", as a JSON format string.  The properties could be anything defined in the schema. For file system schema, it could be location, whether writable, default input format, etc. For other schemas, there would be schema-dependent properties.  In case we add new properties to schema in later release, there is no need to change the implementation of "describe schema".  In other words, "describe schema/database" should return the content shown in webui storage tab for that particular schema.

arina-ielchiieva commented:
@jinfengni, 
I agree that it could be useful to display content shown in webui storage tab as well.
I have added new column `properties` which will contain this information.
I will still keep physical_location in separate column as it's generated from several properties (from FileSystemConfig.connection and WorkspaceConfig.location)  
plus it's was one of initial requirements (https://maprdrill.atlassian.net/browse/MD-603).

jinfengni commented:
@arina-ielchiieva ,
1. the link you posted is actually not a public document. People in the community could not see the content. You may paste some discussion to the public JIRA, if you like.
2. I'm still not clear how you will populate "physical_location" for schema other than file system schema. For instance, HBase, MongoDb, JDBC. 
I understand that Hive will return "location". That makes sense, since Hive only deals with data on file systems. For Drill, it has to connect different storage plugin/schema, including file system schemas. If we put something only works for one type of storage plugin, then we will run into trouble in the future when we have to extend to other storage plugin.

arina-ielchiieva commented:
By default physical location is returned as INFORMATION UNAVAILABLE (as defined in AbstractStoragePlugin.getPhysicalLocation method). When adding new storage plugin, you are not obliged to override it, only if you need it to be overridden as done in FileSystemPlugin.java

jinfengni commented:
I'm not talking about implementation side (AbstractStoragePlugin.getPhysicalLocation etc). Before we go to implementation, we probably want to define what's the expected behavior for "describe schema".  For now, I'm not convinced that "physical_location" column in the output would make sense for non-filesystem schema.

arina-ielchiieva commented:
@jinfengni 
I have updated PR. Now describe command returns only one column `properties` which contains information shown in webui storage tab. Also I have added physicalLocation to workspace config.

jinfengni commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceConfig.java:
```diff
@@ -25,23 +25,27 @@
     this.location = location;
     this.writable = writable;
     this.defaultInputFormat = defaultInputFormat;
+    this.physicalLocation = physicalLocation;
```
WorkspaceConfig already has "location". Why do you prefer adding a new "physicalLocation"? Is it possible to use "location" in stead?

jinfengni commented on exec/java-exec/src/main/resources/bootstrap-storage-plugins.json:
```diff
@@ -6,11 +6,13 @@
           location: "/",
-          writable: false
+          writable: false,
+          physicalLocation: "file:///"
```
Looks like "physicalLocation" could be constructed as "connection" + "location" ?

arina-ielchiieva commented:
@jinfengni 
PR is updated, physical_location is removed.

jinfengni commented:
The new output looks better. 

I checked hive's <describe database> command. Seems it lists schema name and its relevant properties. Does it make sense to add schema | database name in the output?

arina-ielchiieva commented:
@jinfengni, agree. Added schema in the output and updated PR comment with example.

jinfengni commented on exec/java-exec/src/test/java/org/apache/drill/exec/sql/TestInfoSchema.java:
```diff
@@ -351,4 +359,51 @@ public void showFilesWithDefaultSchema() throws Exception{
+        .sqlQuery("describe schema dfs_test")
+        .unOrdered()
+        .baselineColumns("schema", "properties")
+        .baselineValues("dfs_test", properties)
```
I'm not sure if it's a good idea to put the complete "properties" as the expected output in unit test. Doing that makes this unit test prone to failure, in case we change StoragePlugin interface, and add / delete contents in the schema properties. This is similar to Explain Plan output. If we capture everything, it becomes annoying in the future, since any slight change to the planer might fail the unit test.

In stead of capturing everything, can you just verify some "stable" part of properties, like "location" etc, to make this unit test less likely to fail in the future?

arina-ielchiieva replied:
Test has been re-written to verify only stable parts.

jinfengni commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DescribeSchemaHandler.java:
```diff
@@ -0,0 +1,129 @@
+    Object workspaces = configMap.remove("workspaces");
+    if (workspaces != null) {
+      Map map = (Map) workspaces;
+      String key = names.size() > 1 ? names.get(1) : FileSystemSchemaFactory.DEFAULT_WS_NAME;
```
Are you assuming "DESCRIBE schemas" has to applied to full path of schema name?  Can I use the following:

 use dfs;

DESCRIBE schema tmp; // which will get the properties of dfs.tmp

If you see it's not a small effort to add such support, you may put some comment. We may address that later on.

arina-ielchiieva replied:
Yes, I assume we have full path to schema name. 
I will create Jira where we will add support to for
use dfs;
DESCRIBE schema tmp; // which will get the properties of dfs.tmp;
as enhancement to describe schema functionality.

jinfengni commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DescribeSchemaHandler.java:
```diff
@@ -0,0 +1,129 @@
+      try {
+        storagePlugin = context.getStorage().getPlugin(schema.names.get(0));
+      } catch (ExecutionSetupException e) {
+        throw UserException.validationError()
```
Any reason why you treat ExecutionSetupException as UserException:validationError?

arina-ielchiieva replied:
No particular reason. replaced it with DrillRuntimeException (searched through Drill code, mostly DrillRuntimeException is re-thrown instead of ExecutionSetupexception).

jinfengni commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DescribeSchemaHandler.java:
```diff
@@ -0,0 +1,129 @@
+      try {
+        properties = getPropertiesAsJsonString(schema.names, storagePlugin.getConfig());
+      } catch (JsonProcessingException e) {
+        throw UserException.parseError()
```
Same as above.

arina-ielchiieva replied:
Replaced it with DrillRuntimeException.

jinfengni commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/DescribeSchemaHandler.java:
```diff
@@ -0,0 +1,129 @@
+   */
+  private String getPropertiesAsJsonString(List<String> names, StoragePluginConfig config) throws JsonProcessingException {
+    final Map configMap = mapper.convertValue(config, Map.class);
+    Object workspaces = configMap.remove("workspaces");
```
The logic of L104 - 117 seems to apply to FileSystemSP only. Better make sure it will be called only for FilesystemSP (what if other storage plugin config also adds "workspace" later on? )

arina-ielchiieva replied:
Added changes in second commit,

jinfengni commented:
LGTM. 

+1

Thanks for the PR, @arina-ielchiieva !

jinfengni commented:
I'll will regression and merge this PR.
